File: scripts/find_zh_kanji_candidates.py
#!/usr/bin/env python3
"""
简化字・繁體字 -> 日本語漢字 の變換表を育てるための候補抽出。

Unicode 公式の Unihan データベース (kSimplifiedVariant / kTraditionalVariant) を
既存の kyuji_map.json (日本語の新字体<->正字對應) と突き合はせ、
まだ zh_char_map.tsv / zh_compound_map.tsv / zh_ambiguous_characters.json に
載ってゐない文字を、實際に指定したコーパスへ出現する頻度順で報吿する。

候補はあくまで報吿のみ。誤爆(日本語として正規に使はれる字を誤って
變換してしまふこと)を避けるため、必ず人手で個別に検討してから
zh_char_map.tsv 等へ追記すること。

使ひ方:
    python3 scripts/find_zh_kanji_candidates.py \
        --scan-dir wiki

既に取得濟みの Unihan_Variants.txt を使ふ場合:
    python3 scripts/find_zh_kanji_candidates.py \
        --scan-dir wiki --unihan-variants /path/to/Unihan_Variants.txt
"""

from __future__ import annotations

import argparse
import io
import json
import urllib.request
import zipfile
from collections import Counter
from pathlib import Path

UNIHAN_ZIP_URL = "https://www.unicode.org/Public/UCD/latest/ucd/Unihan.zip"


def parse_args() -> argparse.Namespace:
    script_dir = Path(__file__).resolve().parent
    data_dir = script_dir.parent / "dic"
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--scan-dir",
        type=Path,
        action="append",
        required=True,
        help="候補の出現頻度を數へる對象ディレクトリ(複數指定可)",
    )
    parser.add_argument("--data-dir", type=Path, default=data_dir)
    parser.add_argument(
        "--unihan-variants",
        type=Path,
        help="取得濟みの Unihan_Variants.txt を使ふ。未指定なら unicode.org から取得する",
    )
    parser.add_argument(
        "--min-count",
        type=int,
        default=1,
        help="コーパス中の最小出現囘數。既定は1",
    )
    parser.add_argument("--output", type=Path, help="TSV の出力先。未指定なら stdout")
    return parser.parse_args()


def fetch_unihan_variants() -> str:
    request = urllib.request.Request(
        UNIHAN_ZIP_URL, headers={"User-Agent": "Mozilla/5.0"}
    )
    with urllib.request.urlopen(request, timeout=60) as response:
        blob = response.read()
    with zipfile.ZipFile(io.BytesIO(blob)) as archive:
        with archive.open("Unihan_Variants.txt") as fh:
            return fh.read().decode("utf-8")


def to_char(codepoint: str) -> str:
    return chr(int(codepoint[2:], 16))


def load_variant_pairs(variants_text: str) -> dict[str, set[str]]:
    """简化字 -> 對應する繁體字(候補)の集合。

    kTraditionalVariant は「この字は简化字で、繁體字はこちら」を示す。
    kSimplifiedVariant はその逆向き(繁體字 -> 简化字)なので、
    ここでは候補抽出の方向を誤らないやう kTraditionalVariant のみ使ふ。
    """
    pairs: dict[str, set[str]] = {}
    for line in variants_text.splitlines():
        if not line or line.startswith("#"):
            continue
        parts = line.split("\t")
        if len(parts) < 3:
            continue
        src_code, field, values = parts[0], parts[1], parts[2]
        if field != "kTraditionalVariant":
            continue
        src = to_char(src_code)
        for token in values.split():
            code = token.split("<")[0]
            if not code.startswith("U+"):
                continue
            pairs.setdefault(src, set()).add(to_char(code))
    return pairs


def load_known_sources(data_dir: Path) -> set[str]:
    known: set[str] = set()

    kyuji = json.loads((data_dir / "kyuji_map.json").read_text(encoding="utf-8"))
    known.update(kyuji["char_map"].keys())
    known.update(kyuji["char_map"].values())
    known.update(kyuji["ambiguous_characters"].keys())

    for name in ("zh_char_map.tsv", "zh_compound_map.tsv"):
        path = data_dir / name
        if not path.exists():
            continue
        for raw_line in path.read_text(encoding="utf-8").splitlines():
            line = raw_line.strip()
            if not line or line.startswith("#"):
                continue
            known.add(line.split("\t")[0])

    ambiguous_path = data_dir / "zh_ambiguous_characters.json"
    if ambiguous_path.exists():
        known.update(json.loads(ambiguous_path.read_text(encoding="utf-8")).keys())

    return known


def resolve_ja_candidates(
    char: str, variant_pairs: dict[str, set[str]], reverse_kyuji: dict[str, str]
) -> set[str]:
    candidates: set[str] = set()
    for partner in variant_pairs.get(char, set()):
        if partner == char:
            continue
        candidates.add(reverse_kyuji.get(partner, partner))
    return candidates


def count_corpus_chars(scan_dirs: list[Path], targets: set[str]) -> Counter[str]:
    counter: Counter[str] = Counter()
    for scan_dir in scan_dirs:
        for path in scan_dir.rglob("*.md"):
            text = path.read_text(encoding="utf-8", errors="ignore")
            for ch in text:
                if ch in targets:
                    counter[ch] += 1
    return counter


def main() -> int:
    args = parse_args()

    variants_text = (
        args.unihan_variants.read_text(encoding="utf-8")
        if args.unihan_variants
        else fetch_unihan_variants()
    )
    variant_pairs = load_variant_pairs(variants_text)

    kyuji = json.loads((args.data_dir / "kyuji_map.json").read_text(encoding="utf-8"))
    reverse_kyuji: dict[str, str] = {}
    for shinjitai, seiji in kyuji["char_map"].items():
        reverse_kyuji.setdefault(seiji, shinjitai)

    known_sources = load_known_sources(args.data_dir)

    unknown_variant_chars = {c for c in variant_pairs if c not in known_sources}
    counter = count_corpus_chars(args.scan_dir, unknown_variant_chars)

    rows = []
    for char, count in counter.most_common():
        if count < args.min_count:
            continue
        candidates = resolve_ja_candidates(char, variant_pairs, reverse_kyuji)
        candidates.discard(char)
        if not candidates:
            continue
        rows.append((char, count, sorted(candidates)))

    lines = ["# char\tcount\tja_candidates(要人手選別)"]
    for char, count, candidates in rows:
        lines.append(f"{char}\t{count}\t{'|'.join(candidates)}")
    output_text = "\n".join(lines) + "\n"

    if args.output:
        args.output.write_text(output_text, encoding="utf-8")
        print(f"wrote {args.output} ({len(rows)} candidates)")
    else:
        print(output_text, end="")

    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/find_zh_kanji_candidates.clj
#!/usr/bin/env bb
(ns find-zh-kanji-candidates
  "简化字・繁體字から日本語漢字への變換候補を抽出する。"
  (:require
    [babashka.cli :as cli]
    [babashka.fs :as fs]
    [babashka.http-client :as http]
    [cheshire.core :as json]
    [clojure.string :as str])
  (:import
    (java.io
      ByteArrayInputStream)
    (java.nio
      ByteBuffer)
    (java.nio.charset
      CodingErrorAction
      StandardCharsets)
    (java.util.zip
      ZipInputStream)))


(def unihan-zip-url "https://www.unicode.org/Public/UCD/latest/ucd/Unihan.zip")


(def data-dir
  (-> *file*
      fs/path
      fs/parent
      fs/parent
      (fs/path "dic")))


(defn parse-args
  []
  (let [{:keys [opts args]}
        (cli/parse-args
          *command-line-args*
          {:coerce {:data-dir fs/path
                    :min-count parse-long
                    :output fs/path
                    :scan-dir [fs/path]
                    :unihan-variants fs/path}
           :spec {:data-dir {:default data-dir}
                  :min-count {:default 1}
                  :scan-dir {:require true}}})]
    (when (seq args)
      (throw (ex-info "Unexpected positional arguments" {:args args})))
    opts))


(defn decode-utf8-ignoring-errors
  [bytes]
  (let [decoder (doto (.newDecoder StandardCharsets/UTF_8)
                  (.onMalformedInput CodingErrorAction/IGNORE)
                  (.onUnmappableCharacter CodingErrorAction/IGNORE))]
    (str (.decode decoder (ByteBuffer/wrap bytes)))))


(defn fetch-unihan-variants
  []
  (let [{:keys [body status]} (http/get unihan-zip-url {:as :bytes :throw false})]
    (when-not (= status 200)
      (throw (ex-info "Unable to fetch Unihan ZIP" {:status status})))
    (with-open [zip (ZipInputStream. (ByteArrayInputStream. body))]
      (loop [entry (.getNextEntry zip)]
        (cond
          (nil? entry) (throw (ex-info "Unihan_Variants.txt is missing from ZIP" {}))
          (= "Unihan_Variants.txt" (.getName entry))
          (decode-utf8-ignoring-errors (.readAllBytes zip))
          :else (recur (.getNextEntry zip)))))))


(defn to-char
  [codepoint]
  (String. (Character/toChars (Integer/parseInt (subs codepoint 2) 16))))


(defn load-variant-pairs
  [variants-text]
  (reduce (fn [pairs line]
            (let [parts (str/split line #"\t")]
              (if (or (str/blank? line)
                      (str/starts-with? line "#")
                      (< (count parts) 3)
                      (not= "kTraditionalVariant" (nth parts 1)))
                pairs
                (let [source (to-char (nth parts 0))]
                  (reduce (fn [pairs token]
                            (let [code (first (str/split token #"<"))]
                              (if (str/starts-with? code "U+")
                                (update pairs source (fnil conj #{}) (to-char code))
                                pairs)))
                          pairs
                          (str/split (nth parts 2) #"\s+"))))))
          {}
          (str/split-lines variants-text)))


(defn load-known-sources
  [data-dir]
  (let [kiuzi (json/parse-string (slurp (str (fs/path data-dir "kiuzi_map.json"))))
        known (into #{} (keys (get kiuzi "char_map")))]
    (into known
          (concat (vals (get kiuzi "char_map"))
                  (keys (get kiuzi "ambiguous_characters"))
                  (mapcat (fn [name]
                            (let [path (fs/path data-dir name)]
                              (if-not (fs/exists? path)
                                []
                                (for [line (str/split-lines (slurp (str path)))
                                      :let [line (str/trim line)]
                                      :when (and (not (str/blank? line))
                                                 (not (str/starts-with? line "#")))]
                                  (first (str/split line #"\t"))))))
                          ["zh_char_map.tsv" "zh_compound_map.tsv"])
                  (let [path (fs/path data-dir "zh_ambiguous_characters.json")]
                    (if (fs/exists? path)
                      (keys (json/parse-string (slurp (str path))))
                      []))))))


(defn load-compound-sources
  [data-dir]
  (let [path (fs/path data-dir "zh_compound_map.tsv")]
    (if-not (fs/exists? path)
      #{}
      (into #{}
            (for [line (str/split-lines (slurp (str path)))
                  :let [line (str/trim line)
                        source (first (str/split line #"\t"))]
                  :when (and (not (str/blank? line))
                             (not (str/starts-with? line "#"))
                             (not (str/blank? source)))]
              source)))))


(defn code-point-compare
  [left right]
  (let [left-code-points (.toArray (.codePoints ^String left))
        right-code-points (.toArray (.codePoints ^String right))
        limit (min (alength left-code-points) (alength right-code-points))]
    (loop [index 0]
      (if (= index limit)
        (Long/compare (alength left-code-points) (alength right-code-points))
        (let [order (Long/compare (aget left-code-points index)
                                  (aget right-code-points index))]
          (if (zero? order)
            (recur (inc index))
            order))))))


(defn resolve-ja-candidates
  [char variant-pairs reverse-kiuzi]
  (into #{}
        (for [partner (get variant-pairs char #{})
              :when (not= partner char)]
          (get reverse-kiuzi partner partner))))


(defn mark-covered
  [text compound-sources]
  (let [covered (boolean-array (.length text))]
    (doseq [compound compound-sources]
      (loop [start 0]
        (let [index (.indexOf text compound start)]
          (when-not (= -1 index)
            (dotimes [offset (.length compound)]
              (aset-boolean covered (+ index offset) true))
            (recur (+ index (.length compound)))))))
    covered))


(defn count-text-characters
  [state text targets compound-sources]
  (let [covered (mark-covered text compound-sources)]
    (loop [index 0
           state state]
      (if (>= index (.length text))
        state
        (let [code-point (.codePointAt text index)
              char-count (Character/charCount code-point)
              char (String. (Character/toChars code-point))]
          (recur (+ index char-count)
                 (if (or (aget covered index) (not (contains? targets char)))
                   state
                   (let [count (get-in state [:counts char] 0)]
                     (cond-> (update-in state [:counts char] (fnil inc 0))
                       (zero? count)
                       (-> (assoc-in [:first-index char] (:next-index state))
                           (update :next-index inc)))))))))))


(defn count-corpus-characters
  [scan-dirs targets compound-sources]
  (reduce (fn [state scan-dir]
            (reduce (fn [state path]
                      (count-text-characters state
                                             (decode-utf8-ignoring-errors
                                               (fs/read-all-bytes path))
                                             targets
                                             compound-sources))
                    state
                    (concat (fs/glob scan-dir "*.md")
                            (fs/glob scan-dir "**/*.md"))))
          {:counts {} :first-index {} :next-index 0}
          scan-dirs))


(defn ordered-counts
  [{:keys [counts first-index]}]
  (sort (fn [[char-a count-a] [char-b count-b]]
          (let [count-order (compare count-b count-a)]
            (if (zero? count-order)
              (compare (get first-index char-a) (get first-index char-b))
              count-order)))
        counts))


(defn main
  []
  (let [{:keys [data-dir min-count output scan-dir unihan-variants]} (parse-args)
        variants-text (if unihan-variants
                        (decode-utf8-ignoring-errors
                          (fs/read-all-bytes unihan-variants))
                        (fetch-unihan-variants))
        variant-pairs (load-variant-pairs variants-text)
        kiuzi (json/parse-string (slurp (str (fs/path data-dir "kiuzi_map.json"))))
        reverse-kiuzi (reduce (fn [reversed [shinjitai seiji]]
                                (if (contains? reversed seiji)
                                  reversed
                                  (assoc reversed seiji shinjitai)))
                              {}
                              (sort (fn [[left _] [right _]]
                                      (code-point-compare left right))
                                    (get kiuzi "char_map")))
        known-sources (load-known-sources data-dir)
        compound-sources (load-compound-sources data-dir)
        unknown-variant-chars (reduce disj (set (keys variant-pairs)) known-sources)
        counts (count-corpus-characters scan-dir
                                        unknown-variant-chars
                                        compound-sources)
        rows (for [[char count] (ordered-counts counts)
                   :when (>= count min-count)
                   :let [candidates (disj (resolve-ja-candidates char
                                                                 variant-pairs
                                                                 reverse-kiuzi)
                                          char)]
                   :when (seq candidates)]
               (str char "\t" count "\t"
                    (str/join "|" (sort code-point-compare candidates))))
        output-text (str "# char\tcount\tja_candidates(要人手選別)\n"
                         (str/join "\n" rows)
                         "\n")]
    (if output
      (do
        (spit (str output) output-text)
        (println (format "wrote %s (%d candidates)" output (count rows))))
      (print output-text))
    0))


(System/exit (main))
